app/Http/Controllers/SquareController.php

```diff
@@ -0,0 +1,206 @@
+                [
+                    'message' => 'We could not contact Square to begin your payment.',
+                ]
+            );
+        }
+
+        $checkout = $checkoutResponse->getResult()->getCheckout();
+
+        $payment->checkout_id = $checkout->getId();
+        $payment->order_id = $checkout->getOrder()->getId();
+        $payment->save();
+
+        return redirect($checkout->getCheckoutPageUrl());
+    }
+
+    public function complete(SquareCompleteRequest $request)
+    {
+        $payment = Payment::where('id', $request->input('referenceId'))->firstOrFail();
+
+        if ($payment->checkout_id !== $request->input('checkoutId')) {
+            return view(
+                'square.error',
+                [
+                    'message' => 'We could not match your payment in our database.',
+                ]
+            );
+        }
+
+        $square = new SquareClient([
+            'accessToken' => config('square.access_token'),
+            'environment' => config('square.environment'),
+        ]);
+
+        $ordersApi = $square->getOrdersApi();
+
+        $retrieveOrderResponse = $ordersApi->retrieveOrder($payment->order_id);
+
+        if (! $retrieveOrderResponse->isSuccess()) {
+            Log::error(self::class.' Error retrieving order - '.json_encode($retrieveOrderResponse->getErrors()));
+
+            return view(
+                'square.error',
+                [
+                    'message' => 'We could not contact Square to finalize your payment.',
+                ]
+            );
+        }
+
+        $order = $retrieveOrderResponse->getResult()->getOrder();
+
+        switch ($order->getState()) {
+            case OrderState::COMPLETED:
+                $tender = $order->getTenders()[0];
+                $payment->amount = $tender->getAmountMoney()->getAmount() / 100;
+                if (null !== $tender->getProcessingFeeMoney()) {
+                    $payment->processing_fee = $tender->getProcessingFeeMoney()->getAmount() / 100;
+                }
+                $payment->notes = 'Checkout flow completed';
+                $payment->save();
+
+                alert()->success("We've received your payment", 'Success!');
+
+                return redirect('/');
+            case OrderState::CANCELED:
+                return view(
+                    'square.error',
+                    [
+                        'message' => 'Your order was canceled.',
+                    ]
+                );
+            case OrderState::OPEN:
+                return view('square.processing');
+        }
+    }
+
+    private static function generateUniqueId(): string
+    {
+        return bin2hex(openssl_random_pseudo_bytes(32));
+    }
+
+    private static function calculateSurcharge(int $amount): int
+    {
+        return (int) round(
+            (($amount + self::PER_TRANSACTION_FEE) / ((100 - self::PERCENTAGE_FEE) / 100)) - $amount,
+            0,
+            PHP_ROUND_HALF_UP
```

Comment:
This is the specific rounding mode Square uses - https://squareup.com/help/us/en/article/5092-rounding